### PR TITLE
Use JetStream Instead of Regular NATS and Add Metrics

### PR DESCRIPTION
**Pull request by kianaza**
### 1. JetStream Implementation
Created `jetstream.go` which contains functions using `jetstreamContext` for the messaging process.
Retained the existing `nats.go` for regular messaging, which can be deleted upon approval.

### 2. New Metrics
- **Connection Errors**(_Counter_): A counter to track the number of disconnections from NATS.
- **Success Counter**(_CounterVec_): Tracks the number of successful publishes and consumes, as well as publish failures.
- **Latency**(_Histogram_): Measures the latency between publish and consume actions. The time of publishing is sent via the message data, and the consumer calculates the latency.

### CODE REVIEW

**Review by 1995parham — CHANGES_REQUESTED** [review]

File: internal/config/default.go [hunk 1]
```diff
@@ -15,10 +16,17 @@ func Default() Config {
...
+			URL:                    "localhost:4222",
+			PublishInterval:        2 * time.Second,
+			RequestTimeout:         50 * time.Millisecond,
+			DefaultSubject:         "test",
+			MaxPubAcksInflight:     1000,
+			QueueSubscriptionGroup: "group",
+			FlushTimeout:           2 * time.Second,
+		},
+		Metric: metric.Config{
+			Address: ":8080",
```
Comment (sinamna) [inline internal/config/default.go:28]: have a section as metric.server.address. i thing metric config as general may cause confusion

File: internal/natsclient/jetstream.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+func (j *Jetstream) createJetstreamContext() {
+	var err error
+	j.jetstream, err = j.connection.JetStream()
+	if err != nil {
+		j.logger.Panic("could not connect to jetstream", zap.Error(err))
+	}
+}
+
+func (j *Jetstream) createStream() {
+	_, err := j.jetstream.AddStream(&nats.StreamConfig{
```
Comment (sinamna) [inline internal/natsclient/jetstream.go:67]: having these as config parameters is better. some time we dont wanna create stream but we want to add a subject to it

File: internal/natsclient/jetstream.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		Name:     "test",
+		Subjects: []string{"test.*"},
+	})
+	if err != nil {
+		j.logger.Panic("could not add stream", zap.Error(err))
+	}
+}
+
+func (j *Jetstream) StartJetstreamMessaging() {
```
Comment (sinamna) [inline internal/natsclient/jetstream.go:76]: i suggest renaming this method to sth like startblackboxtest or sth.

File: internal/natsclient/jetstream.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	connection *nats.Conn
+	jetstream  nats.JetStreamContext
+	config     *Config
+	logger     *zap.Logger
+	metrics    Metrics
+}
+
+// NewJetstream initializes NATS JetStream connection
+func NewJetstream(config Config, logger *zap.Logger) *Jetstream {
+	j := Jetstream{
```
[the remark this answers is not in the document]
Comment (kianaza) [reply internal/natsclient/jetstream.go]: Thank you for your suggestion. I have applied the fix.

File: internal/natsclient/jetstream.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		j.logger.Panic("could not connect to nats", zap.Error(err))
+	}
+
+	j.connection.SetDisconnectErrHandler(func(_ *nats.Conn, err error) {
+		j.metrics.ConnectionErrors.Add(1)
+		j.logger.Error("nats disconnected", zap.Error(err))
+	})
+
+	j.connection.SetReconnectHandler(func(_ *nats.Conn) {
+		j.logger.Warn("nats reconnected")
```
[the remark this answers is not in the document]
Comment (kianaza) [reply internal/natsclient/jetstream.go:54]: Of course, I have added the reconnection metrics as requested.

File: internal/config/default.go (hunk 1 above)
Comment (kianaza) [reply internal/config/default.go]: Fixed. 👍

File: internal/natsclient/jetstream.go (hunk 2 above)
Comment (kianaza) [reply internal/natsclient/jetstream.go]: Really good point. I have added it to the config.

File: internal/natsclient/jetstream.go (hunk 3 above)
Comment (kianaza) [reply internal/natsclient/jetstream.go]: Fixed. 👍

**Pull request merged**